### PR TITLE
perf(vector): updated marshalling of vector

### PR DESCRIPTION
**Pull request by ghost**
Earlier we were unmarshalling bytes to []float64 by iterating on each element and reading it little endian. But we are now doing it using unsafe pointers. This reduces thee time from O(size(bytes)) to O(1) basically. 
Benchmark stats:
```
goos: linux
goarch: amd64
pkg: github.com/dgraph-io/dgraph/tok/index
cpu: 11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
BenchmarkEncodeDecodeUint64Matrix/Binary_Encoding/Decoding-8              185269              5991 ns/op
BenchmarkEncodeDecodeUint64Matrix/Gob_Encoding/Decoding-8                  46299             25547 ns/op
BenchmarkEncodeDecodeUint64Matrix/JSON_Encoding/Decoding-8                 77666             13723 ns/op
BenchmarkEncodeDecodeUint64Matrix/PB_Encoding/Decoding-8                  500092              2346 ns/op
BenchmarkEncodeDecodeUint64Matrix/Unsafe_Encoding/Decoding-8          1465863          826.2 ns/op
BenchmarkDotProduct/vek:size=96000-8                                      528499              2331 ns/op
BenchmarkDotProduct/dotProduct:size=96000-8                               170630              7765 ns/op
BenchmarkDotProduct/dotProductT:size=96000-8                              145855              8314 ns/op
BenchmarkFloatConverstion/Current:size=96000-8                          284873474                4.172 ns/op
BenchmarkFloatConverstion/pointerFloat:size=96000-8                     263052618                3.988 ns/op
BenchmarkFloatConverstion/littleEndianFloat:size=96000-8                   40446             35287 ns/op

```

Now indexing 500k vectors take about 5 minutes. (more than 5 hours before)

### CODE REVIEW

File: tok/hnsw/helper.go [hunk 1]
```diff
@@ -444,6 +435,68 @@ func (ph *persistentHNSW[T]) createEntryAndStartNodes(
 	return entry, edges, nil
 }
 
+func encodeUint64MatrixUnsafe(matrix [][]uint64) []byte {
```
Comment (mangalaman93) [inline tok/hnsw/helper.go:438]: If we do end up using it, we should draw a diagram of the format for better understanding.

**Review by mangalaman93 — CHANGES_REQUESTED** [review]

File: posting/index.go [hunk 2]
```diff
@@ -649,6 +649,121 @@ type rebuilder struct {
 	fn func(uid uint64, pl *List, txn *Txn) ([]*pb.DirectedEdge, error)
 }
 
+func (r *rebuilder) RunWithoutTemp(ctx context.Context) error {
+	stream := pstore.NewStreamAt(r.startTs)
```
[the remark this answers is not in the document]
Comment (harshil-goel) [reply posting/index.go:653]: It still happens in memory entirely. It's just a side effect that we are updating each thing in temp directory.

File: posting/index.go [hunk 3]
```diff
@@ -649,6 +649,121 @@ type rebuilder struct {
...
+			if found {
+				break
+			}
+		}
+
+		_, err = r.fn(pk.Uid, l, txn)
+		if err != nil {
+			return nil, err
+		}
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (harshil-goel) [reply posting/index.go:734]: Yeah. No issues

File: tok/hnsw/helper.go [hunk 4]
```diff
@@ -335,7 +326,7 @@ func populateEdgeDataFromKeyWithCacheType(
 	if data == nil {
 		return false, nil
 	}
-	err = json.Unmarshal(data.([]byte), &edgeData)
+	err = decodeUint64MatrixUnsafe(data.([]byte), edgeData)
```
[the remark this answers is not in the document]
Comment (harshil-goel) [reply tok/hnsw/helper.go:329]: We have to iterate over the list to make it a protobuff. So to marshal and unmarshal becomes a linear task. Hence taking too much time.

[the remark this answers is not in the document]
Comment (harshil-goel) [reply tok/hnsw/helper.go:307]: BenchmarkEncodeDecodeUint64Matrix/JSON_Encoding/Decoding-8                 91692             12608 ns/op
BenchmarkEncodeDecodeUint64Matrix/PB_Encoding/Decoding-8                  467174              2221 ns/op
BenchmarkEncodeDecodeUint64Matrix/Unsafe_Encoding/Decoding-8             1609965               748.9 ns/op

**Review by mangalaman93 — APPROVED** [review]

**Pull request merged**